### DISCUSSION TITLE
Print to scale with Allmaps?

### DISCUSSION THREAD

**Discussion by sammeltassen** [allmaps/allmaps#468] [General ideas and discussions]
The Allmaps render package supports different renderers: WebGL2Renderer, IntArrayRenderer and the CanvasRenderer.

My students would like to be able to render a collection of maps according to a preset scale, bbox, rotation and ppi, with the aim to print them again on paper. What is the best way to do this?

Let's take [this georeferenced map series of the Netherlands](https://viewer.allmaps.org/?url=https%3A%2F%2Fraw.githubusercontent.com%2Fbmmeijers%2Fiiif-annotations%2Fmain%2Fseries%2Ftmk%2F20231124.json) as an example.

These are the bounds:
```json
[[50.682190099999865, 3.1522464999999995], [53.61182310000017, 7.301122400000001]]
```

I want to make a print on an A1; the dimensions are `594 * 841 mm`. The combined sheets cover an area of `280 * 325 km`. In order to fit all on the A1, the sheets should be printed at scale 1:500.000. This will leave some blank space around the maps.

The preferred PPI for the image is 300 so if the image would fill the entire paper it would measure:

```js
width = Math.round(594 / 25.4 * 300) // 7016 pixels
height = Math.round(841 / 25.4 * 300) // 9933 pixels
```

How to pass this to the renderer? We can calculate the physical dimensions of the maps on paper and the corresponding pixel dimensions. Using the area above:

```js
width = Math.round(280000 / 500000 / 0.0254 * 300) //  6614 pixels
height = Math.round(325000 / 500000 / 0.0254 * 300) // 7677 pixels
```

But this will result in a slightly cropped image. Since the renderer will re-project the map from Bonne (the original projection) to Web Mercator, the maps are distorted and the original proportions are altered, as illustrated below (made with the output of [this notebook](https://observablehq.com/@tudelft/sheet-indices) and the SVG export of [Placemark](https://play.placemark.io/)).

![24-12-06-bonne-mercator-comparison](https://github.com/user-attachments/assets/5f2e7c0c-459f-45bf-96e1-f36eea5c8aec)

As long as the render module doesn't support alternative map projections, we need to select an axis for the scale (and accept the distortions along this axis and incongruity with the other axis). (Of course, distortions are also present in the original maps, but these will be increased by warping to Web Mercator.)

The measured distances of the top and bottom of the Web Mercator bounding box are not the same, as the meridians run parallel in the projection, while the distance between them decreases further from the equator. (The bounding box of the maps in the original Bonne projection would have four *different* geographic coordinates.)

To calculate the correct pixel dimensions we determine the height based on the measured distance of the y-axis, and the width based on the height multiplied by the ratio of the projected coordinates of the bounding box:

```js
// Distance measured with turf.distance
height = Math.round(325863 / 500000 / 0.0254 * 300) //  7698 pixels
// This requires the annotation to have been loaded by the renderer and needs to be combined with the code below
const [xmin, ymin, xmax, ymax] = renderer.warpedMapList.getProjectedBbox()
const ratio = (xmax - xmin) / (ymax - ymin)
width = Math.round(height * ratio) // 6687 pixels
```

Using the CanvasRenderer we can now render the full series in the browser without clipping at the preferred scale:

```js
const canvas = document.getElementById("canvas")
canvas.width = width
canvas.height = height
const renderer = new CanvasRenderer(canvas)
const annotation = await fetch(annoationUrl)
  .then((response) => response.json())
await renderer.addGeoreferenceAnnotation(annotation)
await renderer.render()
```

Using a library like [changeDPI](https://github.com/shutterstock/changeDPI) we can set the correct DPI for the image before downloading. When we place the image in InDesign it will have the right dimensions:

![indesign-screenshot](https://github.com/user-attachments/assets/77ced869-d137-4afc-89c6-1fb9aa1685e9)

You can give it a try yourself in this [Observable Notebook](https://observablehq.com/@allmaps/canvas-renderer) (work in progress...).

Todo:

- Avoid 'page unresponsive' warnings
- Set a custom bounding box
- Define a rotation
- Select the best tile quality
- Rendering in multiple segments
- Describe use cases
- Alternative method using Allmaps CLI and GDAL

**Comment by sammeltassen** [comment]
Alternative method using Allmaps Command Line Interface and GDAL. Step by step:

1. Install Allmaps Command Line Interface

After installing node.js and running `corepack enable` this can be done with pnpm:

`pnpm install -g @allmaps/cli`

Check if this was successful:

`allmaps --help`

You can find the documentation [here](https://github.com/allmaps/allmaps/tree/develop/apps/cli).

2) Install GDAL

[Installation instructions](https://gdal.org/en/latest/download.html#binaries)

Check if this was successful:

`gdal-config--version`

3) Create and run a dezoomify script to download the high res images

First install [dezoomify-rs](https://dezoomify-rs.ophir.dev/), here using [Homebrew](https://brew.sh/):

`brew install dezoomify-rs`

Navigate to a working folder and run the following command. The files will be written in the subdirectory `output/`. 

```bash
curl "https://raw.githubusercontent.com/bmmeijers/iiif-annotations/main/series/tmk/20231124.json" | \
  allmaps annotation image-ids | \
  allmaps script dezoomify -d input | \
  bash
```

For an overview of options run `allmaps script dezoomify --help`. See also the [documentation](https://github.com/allmaps/allmaps/tree/develop/apps/cli#dezoomify) on GitHub.

5) Create and run a GDAL script

We will use the [EPSG:28992](https://epsg.io/28992) Coordinate Reference System, since this is close the the original map projection and will reduce distortions. We also set the transformation type to `thinPlateSpline`.

```bash
curl "https://raw.githubusercontent.com/bmmeijers/iiif-annotations/main/series/tmk/20231124.json" | \
  allmaps script geotiff -s input -d output -c EPSG:28992 -t thinPlateSpline | \
  bash
```

The script will write a bunch of GeoTIFF, GeoJSON en .vrt files in the `output/` folder.

For an overview of options run `allmaps script geotiff --help`. See also the [documentation](https://github.com/allmaps/allmaps/tree/develop/apps/cli#create-geotiffs-using-gdal) on GitHub.

6) Create a composite image with GDAL

Create a composite image by running the following command. We set the target resolution by calculating the amount of meters per pixel based on the scale and PPI:

```js
resolution = 0.0254 * 500000 / 300 // 42.33333333 meters per pixel
```

`gdal_translate output/merged.vrt -tr 42.33333333 42.33333333 merged.tif`

The dimensions of the output file are 6614  x 7643 pixels.

See the [documentation](https://gdal.org/en/latest/programs/gdal_translate.html).

Finally, set the correct DPI of the output image with [ImageMagick](https://imagemagick.org/):

`magick merged.tif -density 300 merged-300dpi.tif`

See the [documentation](https://imagemagick.org/script/command-line-options.php#density).

**Comment by mclaeysb** [comment]
This will be highly improved by #181 

See the updated readme: you can now create a viewport from a size and maps, or from a scale and maps.

From the initial example: if you want to print on a 594 mm by 841 mm paper with ppi 300, you indeed calculate your canvas width and height, create a viewport and render as follows:

```js
// Compute width and height using paper size and ppi
const width = Math.round(594 / 25.4 * 300) // 7016 pixels
const height = Math.round(841 / 25.4 * 300) // 9933 pixels

// Get canvas from a HTML element and size it correctly
const canvas = document.getElementById('canvas')
// Alternatively, create an offscreen canvas:
// const canvas = new OffscreenCanvas(width, height)

// Get canvas context
const context = canvas.getContext('2d')

// Size your canvas correctly
canvas.width = width
canvas.height = height
canvas.style.width = width + 'px'
canvas.style.height = height + 'px'

// Create the renderer and add your annotation
const renderer = new CanvasRenderer(canvas)
await renderer.addGeoreferenceAnnotation(annotation)

// Specify your viewport
const viewport = Viewport.fromSizeAndMaps(
  [width, height], // viewportSize
  renderer.warpedMapList, // all maps in the renderer's warpedMapList will be used to compute a viewport
)

// The scale that will be applied can be read from the viewport
const projectedGeoPerViewportScale = viewport.projectedGeoPerViewportScale

// Render
await renderer.render(viewport)
```

**Comment by sammeltassen** [comment]
Great! Looking forward to try.

For the example, it might be better to set the width and height using the `fromScaleAndMaps` method, since the calculated width and height are based on the original map projection (and this information might not be available). Following the updated README:

```js
// Calculating scale (independent of input map)
printScale = 500000 // 1 to 500000
inchToMeter = 0.0254
pointsPerInch = 300
const renderScale = printScale * inchToMeter / pointsPerInch // 42.33333333 meters per pixel

// Creating a viewport
const viewport = Viewport.fromScaleAndMaps(
  renderScale
  warpedMapList,
)

// Resizing the canvas to the computed viewport
[ canvas.width, canvas.height ] = viewport.canvasSize
```
